features/lorawan/lorastack/mac/LoRaMacChannelPlan.cpp

```diff
@@ -0,0 +1,226 @@
+
+LoRaMacChannelPlan::LoRaMacChannelPlan() : _lora_phy(NULL), _mib(NULL)
+{
+}
+
+LoRaMacChannelPlan::~LoRaMacChannelPlan()
+{
+}
+
+void LoRaMacChannelPlan::activate_channelplan_subsystem(LoRaPHY *phy, LoRaMacMib *mib)
+{
+    _lora_phy = phy;
+    _mib = mib;
+}
+
+lorawan_status_t LoRaMacChannelPlan::set_plan(const lorawan_channelplan_t& plan)
+{
+    channel_params_t mac_layer_ch_params;
+    lorawan_status_t status;
+
+    get_phy_params_t get_phy;
+    phy_param_t phy_param;
+    uint8_t max_num_channels;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = _lora_phy->get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.value;
+
+    // check if user is setting more channels than supported
+    if (plan.nb_channels > max_num_channels) {
+        return LORAWAN_STATUS_PARAMETER_INVALID;
+    }
+
+    for (uint8_t i = 0; i < plan.nb_channels; i++) {
+
+        mac_layer_ch_params.band = plan.channels[i].ch_param.band;
+
+        mac_layer_ch_params.dr_range.fields.max = plan.channels[i].ch_param.dr_range.fields.max;
+        mac_layer_ch_params.dr_range.fields.min = plan.channels[i].ch_param.dr_range.fields.min;
+        mac_layer_ch_params.dr_range.value = plan.channels[i].ch_param.dr_range.value;
+        mac_layer_ch_params.frequency = plan.channels[i].ch_param.frequency;
+        mac_layer_ch_params.rx1_frequency = plan.channels[i].ch_param.rx1_frequency;
+
+        status = _lora_phy->add_channel(&mac_layer_ch_params, plan.channels[i].id);
+
+        if (status != LORAWAN_STATUS_OK) {
+            return status;
+        }
+    }
+
+    return LORAWAN_STATUS_OK;
+}
+
+lorawan_status_t LoRaMacChannelPlan::get_plan(lorawan_channelplan_t& plan,
+                                             loramac_protocol_params *params)
+{
+    if (params == NULL) {
+        return LORAWAN_STATUS_PARAMETER_INVALID;
+    }
+
+    loramac_mib_req_confirm_t mib_confirm;
+    lorawan_status_t status;
+
+    get_phy_params_t get_phy;
+    phy_param_t phy_param;
+    uint8_t max_num_channels;
+    uint16_t *channel_masks;
+    uint8_t count = 0;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = _lora_phy->get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.value;
+
+    // Now check the Default channel mask
+    get_phy.attribute = PHY_CHANNELS_MASK;
+    phy_param = _lora_phy->get_phy_params(&get_phy);
+    channel_masks = phy_param.channel_mask;
+
+    // Request Mib to get channels
+    memset(&mib_confirm, 0, sizeof(mib_confirm));
+    mib_confirm.type = MIB_CHANNELS;
+
+    status = _mib->get_request(&mib_confirm, params);
+
+    if (status != LORAWAN_STATUS_OK) {
+        return status;
+    }
+
+    for (uint8_t i = 0; i < max_num_channels; i++) {
+        // skip the channels which are not enabled
+        if ((channel_masks[0] & (1U << i)) == 0) {
```

Comment:
Can you sort this out in this patch, and any of the original code doing this?

This should be something like `if ((channel_mask[i / 16] & (1U << (i % 16))) == 0)`. I'd define a helper macro/function `MASK_BIT_TEST(mask, bit)` or something.

If you were using 8-bit units, you could use the `bit_test/set/clear` functions at the bottom of https://github.com/ARMmbed/mbed-os/blob/master/features/FEATURE_COMMON_PAL/nanostack-libservice/mbed-client-libservice/common_functions.h - maybe just adapt those.

Reply:
That's a bug actually. Good catch. We actually test for EU band only, where we have only 16 channels so mask[0] is valid. But this is clearly a bug. I will send a patch through.